Add BeSameDateAs

blairconrad opened the pull request:
Fixes #319.

I kept the name suggested in the issue, since you'd not expressed any displeasure.
I'm happy to make any changes you'd like, of course.

blairconrad commented:
Oh, I'm targeting `master` instead of `develop` like it says in [CONTRIBUTING.md](https://github.com/dennisdoomen/fluentassertions/blob/master/CONTRIBUTING.md), because I couldn't find `develop`.
I hope that's alright. I'm tempted to submit a PR against CONTRIBUTING.md, as I fear it's out of date…
And I'm happy to update the docs, but thought it premature at this point.

dennisdoomen commented:
Yeah, you're right. I've recently dropped the `develop` but forgot to update the contribution docs.

dennisdoomen commented on Tests/FluentAssertions.Shared.Specs/DateTimeAssertionSpecs.cs:
```diff
@@ -792,6 +792,47 @@ public void Should_fail_with_descriptive_message_when_asserting_datetime_has_sec
         }
 
+        [TestMethod]
+        public void Should_succeed_when_asserting_datetime_is_on_same_date()
```
Please write the specs using the same layout and naming conventions as https://github.com/dennisdoomen/fluentassertions/blob/master/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33

blairconrad replied:
Happily. Was not sure which convention to follow from `DateTimeAssertionSpecs` and CONTRIBUTING.md referenced a now-extinct file. I'll meet you later for a CONTRIBUTING.md PR.

dennisdoomen commented on Src/Core/Primitives/DateTimeAssertions.cs:
```diff
@@ -435,5 +435,37 @@ public DateTimeRangeAssertions BeLessThan(TimeSpan timeSpan)
+                .BecauseOf(because, reasonArgs)
+                .FailWith("Expected a {context:date and time} with date {0}{reason}, but found a <null> DateTime.", expected);
+
+            if (success)
```
You can avoid the need of a separate `success` variable by chaining assertions like https://github.com/dennisdoomen/fluentassertions/blob/master/Src/Core/Collections/CollectionAssertions.cs#L37

blairconrad replied:
Will do!

dennisdoomen commented:
Done reviewing. Would you be willing to take another look?

blairconrad commented:
Thanks for the comments. I'd be thrilled to take another look.
Oh, while I have you here, do you prefer an additional commit to address issues, or a replacement commit force pushed to the branch (I'd be inclined to go for the latter, but some people like to see the commits match the conversation history, I've found).

blairconrad commented:
Oh, and assuming I get things wrangled to your satisfaction, what's your policy on cleaning up touched files? I could take a crack at improving consistency and applying the changes you've wanted within those two files. (Generally I'd do this as a second commit in the same PR, but am totally flexible.)
Or I could do nothing. Whatever you like!

dennisdoomen commented:
Thanks @blairconrad 

I usually amend and force push my rework into the PR since for small PRs I don't care too much about the rework as long as the final solution is okay.
With respect to cleaning up, I sometimes to that myself, but tend to keep them in a separate commit. It's up to you if you want to spend time on that. 

PS. Thanks for the update of the contribution guidelines.

dennisdoomen commented:
I'll wait for your rework before I tag the 4.2.0 release.

blairconrad commented:
Oh, I'd definitely clean up in a separate commit.
How's about this:

- I'll rework, amend, and force push my change.
- It will be fabulous, and you'll be very happy.
- Then you can tag the release.
- We'll consider the cleanup/tidy when there's no release pressure. Which still might be tonight. Or tomorrow or something.

I _should_ be able to carve out time for the rework tonight (I'm in UTC-5).

And I'm happy to have updated the contribution guidelines. Everyone wins.

dennisdoomen commented:
:+1:

blairconrad commented:
I think I've addressed all your concerns. But names are hard (for me). Please complain about the spec names if they are inadequate.
Oh. I was not sure I understood `Then`, so I snuck in a bonus nullable DateTime spec without consultation. I hope that's okay.
If not, I will happily remove.

lukefan6 commented:
Hi I just wanna say thank you all for making my issue / request resolved.
:+1:  :+1:  :+1:  :100: 
Looking forward to the next release :smile:

dennisdoomen commented:
It's released: http://www.nuget.org/packages/FluentAssertions/4.2.0

blairconrad commented:
Thanks, @dennisdoomen.
@lukefan6, you're welcome. Enjoy.
